### PR TITLE
SSD1306 Rewrite for Sol keyboard

### PR DESCRIPTION
**Pull request by XScorpion2**
Ground up rewrite of the ssd1306 source using i2c_master instead of i2c. Initial implementation for the Sol keyboard, will expand with more features and keyboard support if it doesn't already work.

## Types of changes

- [ ] Core
- [x] Bugfix
- [ ] New Feature
- [x] Enhancement/Optimization
- [x] Keyboard (addition or update)
- [x] Keymap/Layout/Userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by this PR

* 

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document. (https://docs.qmk.fm/#/contributing)
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/sol/common/i2c_master.c [hunk 1]
```diff
@@ -0,0 +1,223 @@
+/* Library made by: g4lvanix
```
Comment (drashna) [inline keyboards/sol/common/i2c_master.c]: This file should NOT be copied here.   

There is a copy of i2c_master.*  in /drivers/(avr/arm)/, and that is what should be used. 
If that doesn't work, it should be updated to work with your code, while ensuring compatibility with existing code.  
At least, if possible.   

That way, we avoid the code fragmentation issue that we see with various aspects of the code, already.

File: keyboards/sol/common/i2c_master.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
+/* Library made by: g4lvanix
```
Comment (drashna) [inline keyboards/sol/common/i2c_master.h]: Same as the `c` file.

File: keyboards/sol/common/i2c_master.c (hunk 1 above)
Comment (XScorpion2) [reply keyboards/sol/common/i2c_master.c]: makes sense, is that directory part of the standard includes search path?

**Comment by XScorpion2** [comment]
Will check out the oled driver in a bit. Kinda tired after fighting the original i2c and ssd1306 source and need a breather heh.

**Comment by drashna** [comment]
No worries.  I totally understand.

**Comment by XScorpion2** [comment]
So I tried the qwiic oled driver and here are my thoughts and how it stacks up vs the original ssd1306 source and this new one I wrote:

Size free after compiling with qwicc: 4966 bytes free. Compared to the old source at 4530 bytes free, this is really nice considering its feature set of drawing, inverting etc. Though compared it mine at 6496 bytes free, that's quite a large difference. I'd be interested to see how they would compare if I added drawing support, which I was planning on doing eventually.

Ease of switching to the current qwicc: painful. IT has not been fully converted for QMK use from the looks of it as there are mixed up type usages (bool vs uint8_t for example) which prevent it from compiling, redefines of macros that are defined in other files it includes, and incorrect i2c_master source usage in several places. Once I got those problems solved and compiling with switching my keymap / config over to use it, the keyboard bricks itself and becomes an unrecognized usb device. So it's writing to memory out of bounds and crashing the firmware or something unsavory.

Honestly, I wouldn't use it as it is right now. Especially when you compare it to a very well written similar version such as the [Adafruit](https://github.com/adafruit/Adafruit_SSD1306/blob/master/Adafruit_SSD1306.cpp) version which definitely was part of my research for my rewrite.

**Comment by XScorpion2** [comment]
So far still rock solid, minor issue I needed to fix with block based updating: Turns out I really should check the results of the oled col/page command

**Comment by XScorpion2** [comment]
working on compatibility defines to make it easier to switch over to the new source

File: keyboards/sol/common/i2c_master.c (hunk 1 above)
Comment (XScorpion2) [reply keyboards/sol/common/i2c_master.c]: fixed

File: keyboards/sol/common/i2c_master.h (hunk 2 above)
Comment (XScorpion2) [reply keyboards/sol/common/i2c_master.h]: fixed

**Comment by XScorpion2** [comment]
Ok, compatibility defines are done and work with 1 exception: CharacterMatrix... I could implement a dummy struct to take it's place, but that's just stupid and a lot of wasted space. So folks will need to update their code and remove their usage / references to CharacterMatrix, the rest of the APIs can remain and work identically to how they work today with improved rendering performance!

**Comment by XScorpion2** [comment]
Still trying to think how to handle drawing as that then opens up high possibility of needing to write to the same chunk multiple times to get the desired results. This makes it hard to use a chunk based dirty flag as you need some kind of batch system to check the chunk before a set of writes and then after once the batch of writes are complete. Still thinking about how to handle this.

**Comment by XScorpion2** [comment]
@jackhumbert @skullydazed @fredizzimo Any thoughts on this pr? I could take it further if you want and replace all ssd1306 usage with this version updating any code / compile errors as necessary if you think it's worth standardizing all of them.

**Comment by XScorpion2** [comment]
Closed, next iteration: https://github.com/qmk/qmk_firmware/pull/5151

**Pull request closed without merging**